### PR TITLE
#5 Java Version

### PR DESCRIPTION
**Pull request by original-brownbear**
missing:
- [x] Rspec has one test failing since you cannot stub non-existing methods on Java objects
- [x] Rename a few things that now have `_java` in the name
- [x] Cleanup one todo on the correct formatting of the `ua` string (minor_patch field on UA)
- [x] Update Build instructions in Readme
- [x] Not commit `.yaml` files here and use remote resources instead
=> not much left :)

### CODE REVIEW

**Comment by suyograo** [comment]
Nice one @original-brownbear. Do you want to go `java` and `gradle` all the way? I recently converted the GeoIP Filter into java which now works well with `rspec`, `rake` and `gradle`.

The commit is https://github.com/logstash-plugins/logstash-filter-geoip/commit/ec18789d42302eebb3b775ad5c49fb9aba394d40.

So, we basically use [gradle wrapper](https://github.com/logstash-plugins/logstash-filter-geoip/tree/master/gradle/wrapper) and then use the Logstash jars to [compile](https://github.com/logstash-plugins/logstash-filter-geoip/blob/master/build.gradle#L60).

The workflow is better seen here in [.travis.yml](https://github.com/logstash-plugins/logstash-filter-geoip/blob/master/.travis.yml#L12)

**Comment by original-brownbear** [comment]
@suyograo yea I was going for very close to what you linked to. I'll adjust this to be exactly the same in workflow then so we get consistency :)

**Comment by suyograo** [comment]
@original-brownbear assign this for me to review when you are ready.

**Comment by original-brownbear** [comment]
@suyograo will do (and hopefully within the next 45 min too ...) :)

**Comment by original-brownbear** [comment]
@suyograo feel free to give it a look (don't have rights here to assign you outright :)).

One open question for you here that I have right now is:
* What do I do about license headers in files I took from https://github.com/ua-parser/uap-java?

**Comment by suyograo** [comment]
@original-brownbear I added you to `logstash-plugins` org. You should have owner permissions now. Can you append the [contents](https://github.com/ua-parser/uap-java/blob/master/LICENSE) of UAP License file to https://github.com/logstash-plugins/logstash-filter-useragent/blob/master/LICENSE and https://github.com/logstash-plugins/logstash-filter-useragent/blob/master/NOTICE.txt?

**Comment by original-brownbear** [comment]
@suyograo thanks for the permissions :) Should I remove the license headers from the individual `.java` files then (after adding them to the two other files)?

File: logstash-filter-useragent.gemspec [hunk 1]
```diff
@@ -8,10 +8,10 @@ Gem::Specification.new do |s|
   s.authors         = ["Elastic"]
   s.email           = 'info@elastic.co'
   s.homepage        = "http://www.elastic.co/guide/en/logstash/current/index.html"
-  s.require_paths = ["lib"]
+  s.require_paths   = ["lib", Dir.pwd + "/ua-parser/build/libs"]
```
Comment (suyograo) [inline logstash-filter-useragent.gemspec]: We try to use `vendor/jar-dependencies` in all other plugins. Can we change the location of libs to be that? You can copy the build libs to `vendor` using https://github.com/logstash-plugins/logstash-filter-geoip/blob/master/build.gradle#L108

Comment (suyograo) [inline logstash-filter-useragent.gemspec]: You can then just do: `s.require_paths = ["lib", "vendor/jar-dependencies"]`

File: logstash-filter-useragent.gemspec [hunk 2]
```diff
@@ -8,10 +8,10 @@ Gem::Specification.new do |s|
...
 
   # Files
-  s.files = Dir["lib/**/*","spec/**/*","*.gemspec","*.md","CONTRIBUTORS","Gemfile","LICENSE","NOTICE.TXT", "vendor/jar-dependencies/**/*.jar", "vendor/jar-dependencies/**/*.rb", "VERSION", "docs/**/*"]
```
Comment (suyograo) [inline logstash-filter-useragent.gemspec]: See above comment. You can then add this back?

**Comment by suyograo** [comment]
> Should I remove the license headers from the individual .java files then (after adding them to the two other files)?

No, we can leave it as is for now.

File: lib/logstash/filters/useragent.rb [hunk 3]
```diff
@@ -118,48 +108,42 @@ def filter(event)
...
+    @parser.parse(useragent)
   end
 
   private
 
   def set_fields(event, ua_data)
     # UserAgentParser outputs as US-ASCII.
 
-    event.set(@prefixed_name, ua_data.name.dup.force_encoding(Encoding::UTF_8))
+    event.set(@prefixed_name, ua_data.userAgent.family.dup.force_encoding(Encoding::UTF_8))
```
Comment (original-brownbear) [inline lib/logstash/filters/useragent.rb]: @suyograo yea moving that to Java would be great I think. That should save a visible amount of memory :) Should I do it here in this PR?

File: logstash-filter-useragent.gemspec (hunk 1 above)
Comment (original-brownbear) [inline logstash-filter-useragent.gemspec]: @suyograo sure will revert :)

File: logstash-filter-useragent.gemspec (hunk 2 above)
Comment (original-brownbear) [inline logstash-filter-useragent.gemspec]: @suyograo jup will do :)

File: lib/benchmark/cli_transcoder.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    test_case['user_agent_string']
+  end
+end
+
+java_filter = LogStash::Filters::UserAgent.new('source' => 'foo')
+java_filter.register
+
+puts 'Raw Performance with 100% Cache Misses'
+
+start = Time.now
```
Comment (original-brownbear) [inline lib/benchmark/cli_transcoder.rb]: @suyograo maybe add a separate issue for that? Could even drop the benchmark here (as it's really dirty and was just a tool for me ... though it does produce stable numbers :D) and handle it separately?

**Comment by suyograo** [comment]
@original-brownbear I am wondering if we should add a fully Java based `org.logstash.filters.UserAgentFilter` class that does everything inside of that class. Then, useragent.rb will only have `register` and `filter(event)`

 ```
def filter(event)
   if @java_filter.handle(event)
      filter_matched(event)
   end
end
```

where `java_filter` is instance of `org.logstash.filters.UserAgentFilter` which internally has `UAParser` class. 

See https://github.com/logstash-plugins/logstash-filter-geoip/blob/master/src/main/java/org/logstash/filters/GeoIPFilter.java#L98 for example.

WDYT?

File: ua-parser/README.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
+ua_parser Java Library
+======================
+
+This is the Java implementation of [ua-parser](https://github.com/tobie/ua-parser).
```
Comment (jordansissel) [inline ua-parser/README.md]: Curious. Why is the source from a 3rd party library copied here? For other dependencies, we use maven or rubygems, so can you tell me a story about why the ua-parser java source code is included here?

File: lib/benchmark/cli_transcoder.rb (hunk 4 above)
Comment (suyograo) [inline lib/benchmark/cli_transcoder.rb]: makes sense to do in a separate PR.

File: ua-parser/README.md (hunk 5 above)
Comment (original-brownbear) [inline ua-parser/README.md]: @jordansissel the code here is a heavily cleaned up (optimized for less GC mostly) version of the original codebase of `ua-java`. Also, we needed some outright adjustments for it to work with our Ruby API.

**Comment by original-brownbear** [comment]
@suyograo totally agree, moving it all to Java would be nice. But it may make sense to move that to the next step too, so that we at least get one version out that runs from Java but still passes Ruby specs.
Otherwise, we have one step that adjusts RSpec and Java code in one go and lose even the last remnants of traceability? :)

**Comment by suyograo** [comment]
@original-brownbear sounds good. We can do it separately. I am good with current scope.

File: lib/logstash/filters/useragent.rb (hunk 3 above)
Comment (suyograo) [inline lib/logstash/filters/useragent.rb]: As discussed elsewhere, let's defer to a separate PR.

**Comment by suyograo** [comment]
@original-brownbear can we move the UAParser src code from top level directory to `src/main/java/org/logstash/filters/uaparser/` ? This will be consistent with what we do in other plugins/core code.

**Comment by original-brownbear** [comment]
@suyograo I think I addressed all points :) (note to self: squash :D)

**Comment by suyograo** [comment]
@original-brownbear when squashing, can you squash to 2 commits: one `added uap java code` which brings in the UAP library, and then second with all our changes?

**Comment by suyograo** [comment]
The code looks good and all tests are passing. I sent you an email about headers. Once you change that, I'll do a final review and you can squash and merge.

**Comment by original-brownbear** [comment]
@suyograo sure the squash thing is trivial, email is ack :)

**Comment by original-brownbear** [comment]
@suyograo licenses adjusted according :)

**Comment by suyograo** [comment]
LGTM. Please squash as we spoke above and merge.

**Comment by suyograo** [comment]
@original-brownbear here's another GitHub tip you can use: https://help.github.com/articles/closing-issues-via-commit-messages/

So you can do `Fixes #5` in the last commit and it will auto close #5 with this PR trail.

**Comment by original-brownbear** [comment]
@suyograo on it :) Let's see how cleanly I can get the file copying separated out :)

**Comment by original-brownbear** [comment]
@suyograo squashed -> mergin :)

**Comment by original-brownbear** [comment]
@suyograo btw. remaining work here hasn't been forgotten. Will add issues for that in a bit :)

**Comment by suyograo** [comment]
@original-brownbear to publish these awesome changes, can you bump version here in a separete PR: https://github.com/logstash-plugins/logstash-filter-useragent/blob/master/logstash-filter-useragent.gemspec#L4 to 3.1.0

and add an entry in changelog? https://github.com/logstash-plugins/logstash-filter-useragent/blob/master/CHANGELOG.md

Changelog is user facing; we can omit the implementation details and say we used a Java library to improve performance by 2.5x. I would also note the change in the cache size default and mention it wouldn't put too much memory pressure :)

**Comment by original-brownbear** [comment]
@suyograo sure will do in ~1h :)

**Comment by original-brownbear** [comment]
@suyograo there we go: https://github.com/logstash-plugins/logstash-filter-useragent/pull/40 :)
